File: internal/storage/localfs/localfs.go
package localfs

import (
	"context"
	"errors"
	"io/fs"
	"os"
	"path/filepath"
	"sort"
	"strings"

	"github.com/sean/s3-message-service/internal/storage"
)

type Store struct {
	root string
}

func New(root string) (*Store, error) {
	if strings.TrimSpace(root) == "" {
		return nil, storage.ErrInvalidObjectKey
	}
	if err := os.MkdirAll(root, 0o755); err != nil {
		return nil, err
	}
	return &Store{root: root}, nil
}

func (store *Store) Put(ctx context.Context, key string, data []byte, options storage.PutOptions) error {
	if err := ctx.Err(); err != nil {
		return err
	}
	path, err := store.pathForKey(key)
	if err != nil {
		return err
	}
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		return err
	}

	flags := os.O_WRONLY | os.O_CREATE
	if options.CreateOnly {
		flags |= os.O_EXCL
	} else {
		flags |= os.O_TRUNC
	}

	file, err := os.OpenFile(path, flags, 0o644)
	if err != nil {
		if errors.Is(err, os.ErrExist) {
			return storage.ErrObjectAlreadyExists
		}
		return err
	}
	defer file.Close()

	_, err = file.Write(data)
	return err
}

func (store *Store) Get(ctx context.Context, key string) ([]byte, error) {
	if err := ctx.Err(); err != nil {
		return nil, err
	}
	path, err := store.pathForKey(key)
	if err != nil {
		return nil, err
	}
	data, err := os.ReadFile(path)
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return nil, storage.ErrObjectNotFound
		}
		return nil, err
	}
	return data, nil
}

func (store *Store) Head(ctx context.Context, key string) (storage.ObjectInfo, error) {
	if err := ctx.Err(); err != nil {
		return storage.ObjectInfo{}, err
	}
	path, err := store.pathForKey(key)
	if err != nil {
		return storage.ObjectInfo{}, err
	}
	info, err := os.Stat(path)
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return storage.ObjectInfo{}, storage.ErrObjectNotFound
		}
		return storage.ObjectInfo{}, err
	}
	return storage.ObjectInfo{
		Key:        key,
		Size:       info.Size(),
		ModifiedAt: info.ModTime().UTC(),
	}, nil
}

func (store *Store) List(ctx context.Context, input storage.ListInput) (storage.ListPage, error) {
	if err := ctx.Err(); err != nil {
		return storage.ListPage{}, err
	}
	if input.Limit <= 0 {
		input.Limit = 100
	}
	prefix := strings.TrimLeft(input.Prefix, "/")
	var objects []storage.ListedObject
	walkRoot := store.root
	if prefix != "" {
		prefixDirectory := prefix
		if !strings.HasSuffix(prefixDirectory, "/") {
			lastSlash := strings.LastIndex(prefixDirectory, "/")
			if lastSlash >= 0 {
				prefixDirectory = prefixDirectory[:lastSlash+1]
			} else {
				prefixDirectory = ""
			}
		}
		if prefixDirectory != "" {
			candidate, err := store.pathForKey(prefixDirectory)
			if err != nil {
				return storage.ListPage{}, err
			}
			if _, err := os.Stat(candidate); err != nil {
				if errors.Is(err, os.ErrNotExist) {
					return storage.ListPage{}, nil
				}
				return storage.ListPage{}, err
			}
			walkRoot = candidate
		}
	}

	err := filepath.WalkDir(walkRoot, func(path string, entry fs.DirEntry, walkErr error) error {
		if walkErr != nil {
			return walkErr
		}
		if err := ctx.Err(); err != nil {
			return err
		}
		if entry.IsDir() {
			return nil
		}
		relative, err := filepath.Rel(store.root, path)
		if err != nil {
			return err
		}
		key := filepath.ToSlash(relative)
		if !strings.HasPrefix(key, prefix) {
			return nil
		}
		if input.StartAfter != "" && key <= input.StartAfter {
			return nil
		}
		info, err := entry.Info()
		if err != nil {
			return err
		}
		objects = append(objects, storage.ListedObject{
			Key:        key,
			Size:       info.Size(),
			ModifiedAt: info.ModTime().UTC(),
		})
		return nil
	})
	if err != nil {
		return storage.ListPage{}, err
	}

	sort.Slice(objects, func(left, right int) bool {
		return objects[left].Key < objects[right].Key
	})

	hasMore := false
	if len(objects) > input.Limit {
		hasMore = true
		objects = objects[:input.Limit]
	}
	nextAfterKey := ""
	if len(objects) > 0 {
		nextAfterKey = objects[len(objects)-1].Key
	}

	return storage.ListPage{
		Objects:      objects,
		HasMore:      hasMore,
		NextAfterKey: nextAfterKey,
	}, nil
}

func (store *Store) Delete(ctx context.Context, key string) error {
	if err := ctx.Err(); err != nil {
		return err
	}
	path, err := store.pathForKey(key)
	if err != nil {
		return err
	}
	if err := os.Remove(path); err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return storage.ErrObjectNotFound
		}
		return err
	}
	return nil
}

func (store *Store) pathForKey(key string) (string, error) {
	cleanKey := strings.TrimLeft(key, "/")
	if cleanKey == "" || strings.Contains(cleanKey, "\\") {
		return "", storage.ErrInvalidObjectKey
	}
	cleanPath := filepath.Clean(cleanKey)
	if cleanPath == "." || strings.HasPrefix(cleanPath, "..") || filepath.IsAbs(cleanPath) {
		return "", storage.ErrInvalidObjectKey
	}
	return filepath.Join(store.root, filepath.FromSlash(cleanKey)), nil
}

File: src/storage/localfs.rs
use std::fs::{self, OpenOptions};
use std::io::Write;
use std::path::{Component, Path, PathBuf};

use async_trait::async_trait;
use chrono::{DateTime, Utc};
use walkdir::WalkDir;

use crate::error::{Result, ServiceError};
use crate::storage::{ListInput, ListPage, ListedObject, ObjectInfo, ObjectStore, PutOptions};

#[derive(Clone, Debug)]
pub struct LocalFileStore {
    root: PathBuf,
}

impl LocalFileStore {
    pub fn new(root: impl AsRef<Path>) -> Result<Self> {
        let root = root.as_ref();
        if root.as_os_str().is_empty() {
            return Err(ServiceError::InvalidObjectKey);
        }
        fs::create_dir_all(root)?;
        Ok(Self {
            root: root.to_path_buf(),
        })
    }

    fn path_for_key(&self, key: &str) -> Result<PathBuf> {
        let clean_key = key.trim_start_matches('/');
        if clean_key.is_empty() || clean_key.contains('\\') {
            return Err(ServiceError::InvalidObjectKey);
        }
        let relative = Path::new(clean_key);
        for component in relative.components() {
            match component {
                Component::Normal(_) => {}
                _ => return Err(ServiceError::InvalidObjectKey),
            }
        }
        Ok(self.root.join(relative))
    }
}

#[async_trait]
impl ObjectStore for LocalFileStore {
    async fn put(&self, key: &str, data: &[u8], options: PutOptions) -> Result<()> {
        let path = self.path_for_key(key)?;
        if let Some(parent) = path.parent() {
            fs::create_dir_all(parent)?;
        }

        let mut open_options = OpenOptions::new();
        open_options.write(true);
        if options.create_only {
            open_options.create_new(true);
        } else {
            open_options.create(true).truncate(true);
        }

        let mut file = open_options.open(path).map_err(|error| {
            if error.kind() == std::io::ErrorKind::AlreadyExists {
                ServiceError::ObjectAlreadyExists
            } else {
                ServiceError::Io(error)
            }
        })?;
        file.write_all(data)?;
        Ok(())
    }

    async fn get(&self, key: &str) -> Result<Vec<u8>> {
        let path = self.path_for_key(key)?;
        fs::read(path).map_err(|error| {
            if error.kind() == std::io::ErrorKind::NotFound {
                ServiceError::ObjectNotFound
            } else {
                ServiceError::Io(error)
            }
        })
    }

    async fn head(&self, key: &str) -> Result<ObjectInfo> {
        let path = self.path_for_key(key)?;
        let metadata = fs::metadata(path).map_err(|error| {
            if error.kind() == std::io::ErrorKind::NotFound {
                ServiceError::ObjectNotFound
            } else {
                ServiceError::Io(error)
            }
        })?;
        Ok(ObjectInfo {
            key: key.to_string(),
            size: metadata.len(),
            content_type: String::new(),
            modified_at: DateTime::<Utc>::from(metadata.modified()?),
        })
    }

    async fn list(&self, mut input: ListInput) -> Result<ListPage> {
        if input.limit == 0 {
            input.limit = 100;
        }
        let prefix = input.prefix.trim_start_matches('/').to_string();
        let mut walk_root = self.root.clone();
        if !prefix.is_empty() {
            let mut prefix_directory = prefix.clone();
            if !prefix_directory.ends_with('/') {
                if let Some(last_slash) = prefix_directory.rfind('/') {
                    prefix_directory.truncate(last_slash + 1);
                } else {
                    prefix_directory.clear();
                }
            }
            if !prefix_directory.is_empty() {
                let candidate = self.path_for_key(&prefix_directory)?;
                if !candidate.exists() {
                    return Ok(ListPage::default());
                }
                walk_root = candidate;
            }
        }

        let mut objects = Vec::new();
        for entry in WalkDir::new(&walk_root) {
            let entry = entry.map_err(|error| ServiceError::Storage(error.to_string()))?;
            if !entry.file_type().is_file() {
                continue;
            }
            let relative = entry
                .path()
                .strip_prefix(&self.root)
                .map_err(|error| ServiceError::Storage(error.to_string()))?;
            let key = relative
                .to_string_lossy()
                .replace(std::path::MAIN_SEPARATOR, "/");
            if !key.starts_with(&prefix) {
                continue;
            }
            if !input.start_after.is_empty() && key <= input.start_after {
                continue;
            }
            let metadata = entry
                .metadata()
                .map_err(|error| ServiceError::Io(error.into()))?;
            objects.push(ListedObject {
                key,
                size: metadata.len(),
                modified_at: DateTime::<Utc>::from(metadata.modified()?),
            });
        }

        objects.sort_by(|left, right| left.key.cmp(&right.key));
        let has_more = objects.len() > input.limit;
        if has_more {
            objects.truncate(input.limit);
        }
        let next_after_key = objects
            .last()
            .map(|object| object.key.clone())
            .unwrap_or_default();

        Ok(ListPage {
            objects,
            has_more,
            next_after_key,
        })
    }

    async fn delete(&self, key: &str) -> Result<()> {
        let path = self.path_for_key(key)?;
        fs::remove_file(path).map_err(|error| {
            if error.kind() == std::io::ErrorKind::NotFound {
                ServiceError::ObjectNotFound
            } else {
                ServiceError::Io(error)
            }
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn put_get_list_and_create_only() {
        let temp_dir = tempfile::tempdir().expect("tempdir");
        let store = LocalFileStore::new(temp_dir.path()).expect("store");

        store
            .put(
                "a/b/one.json",
                br#"{"one":1}"#,
                PutOptions {
                    create_only: true,
                    content_type: String::new(),
                },
            )
            .await
            .expect("put one");
        store
            .put(
                "a/b/two.json",
                br#"{"two":2}"#,
                PutOptions {
                    create_only: true,
                    content_type: String::new(),
                },
            )
            .await
            .expect("put two");
        let conflict = store
            .put(
                "a/b/one.json",
                b"{}",
                PutOptions {
                    create_only: true,
                    content_type: String::new(),
                },
            )
            .await
            .expect_err("conflict");
        assert!(matches!(conflict, ServiceError::ObjectAlreadyExists));

        let data = store.get("a/b/one.json").await.expect("get");
        assert_eq!(data, br#"{"one":1}"#);

        let page = store
            .list(ListInput {
                prefix: "a/b/".to_string(),
                start_after: String::new(),
                limit: 1,
            })
            .await
            .expect("list");
        assert_eq!(page.objects.len(), 1);
        assert!(page.has_more);
        assert!(!page.next_after_key.is_empty());

        let next_page = store
            .list(ListInput {
                prefix: "a/b/".to_string(),
                start_after: page.next_after_key,
                limit: 10,
            })
            .await
            .expect("list next");
        assert_eq!(next_page.objects.len(), 1);
        assert_eq!(next_page.objects[0].key, "a/b/two.json");
    }
}
